### PR TITLE
Fix panic in dagreader with raw nodes

### LINKED ISSUE #3399: Panic in gateway trying to access raw leaf node
```
runtime error: invalid memory address or nil pointer dereference gateway_handler.go:81
goroutine 2784 [running]:
runtime/debug.Stack(0x40, 0xc4228130a8, 0x6505f7)
	/usr/lib/go/src/runtime/debug/stack.go:24 +0x79
runtime/debug.PrintStack()
	/usr/lib/go/src/runtime/debug/stack.go:16 +0x22
github.com/ipfs/go-ipfs/core/corehttp.(*gatewayHandler).ServeHTTP.func2()
	/home/kubuxu/AUR/go-ipfs-git/src/src/github.com/ipfs/go-ipfs/core/corehttp/gateway_handler.go:82 +0x160
panic(0xc01020, 0xc420010140)
	/usr/lib/go/src/runtime/panic.go:458 +0x243
github.com/ipfs/go-ipfs/unixfs/io.(*DagReader).Close(0xc422463650, 0xc422ea6284, 0x37)
	/home/kubuxu/AUR/go-ipfs-git/src/src/github.com/ipfs/go-ipfs/unixfs/io/dagreader.go:230 +0x18
panic(0xc01020, 0xc420010140)
	/usr/lib/go/src/runtime/panic.go:458 +0x243
github.com/ipfs/go-ipfs/unixfs/io.(*DagReader).Seek(0xc422463650, 0x0, 0x0, 0xd009b2, 0x19, 0x4)
	/home/kubuxu/AUR/go-ipfs-git/src/src/github.com/ipfs/go-ipfs/unixfs/io/dagreader.go:254 +0x62
net/http.serveContent(0x1138660, 0xc4201122d0, 0xc421eb3860, 0xc422ea628a, 0x31, 0xe7791f701, 0xc400000000, 0x121e380, 0xc422f096a0, 0x7fb438b7e540, ...)
	/usr/lib/go/src/net/http/fs.go:169 +0xc84
net/http.ServeContent(0x1138660, 0xc4201122d0, 0xc421eb3860, 0xc422ea628a, 0x31, 0xe7791f701, 0x0, 0x121e380, 0x7fb438b7e540, 0xc422463650)
	/usr/lib/go/src/net/http/fs.go:134 +0xde
github.com/ipfs/go-ipfs/core/corehttp.(*gatewayHandler).getOrHeadHandler(0xc42204ac40, 0x113a560, 0xc422f58300, 0x1138660, 0xc4201122d0, 0xc421eb3860)
	/home/kubuxu/AUR/go-ipfs-git/src/src/github.com/ipfs/go-ipfs/core/corehttp/gateway_handler.go:230 +0x146c
github.com/ipfs/go-ipfs/core/corehttp.(*gatewayHandler).ServeHTTP(0xc42204ac40, 0x1138660, 0xc4201122d0, 0xc421eb3860)
	/home/kubuxu/AUR/go-ipfs-git/src/src/github.com/ipfs/go-ipfs/core/corehttp/gateway_handler.go:101 +0x5d2
net/http.(*ServeMux).ServeHTTP(0xc4222883c0, 0x1138660, 0xc4201122d0, 0xc421eb3860)
	/usr/lib/go/src/net/http/server.go:2022 +0x7f
github.com/ipfs/go-ipfs/core/corehttp.IPNSHostnameOption.func1.1(0x1138660, 0xc4201122d0, 0xc421eb3860)
	/home/kubuxu/AUR/go-ipfs-git/src/src/github.com/ipfs/go-ipfs/core/corehttp/ipns_hostname.go:31 +0x142
net/http.HandlerFunc.ServeHTTP(0xc42204dca0, 0x1138660, 0xc4201122d0, 0xc421eb3860)
	/usr/lib/go/src/net/http/server.go:1726 +0x44
net/http.(*ServeMux).ServeHTTP(0xc421fff830, 0x1138660, 0xc4201122d0, 0xc421eb3860)
	/usr/lib/go/src/net/http/server.go:2022 +0x7f
net/http.(Handler).ServeHTTP-fm(0x1138660, 0xc4201122d0, 0xc421eb3860)
	/usr/lib/go/src/net/http/h2_bundle.go:4084 +0x4d
gx/ipfs/QmR3KwhXCRLTNZB59vELb2HhEWrGy9nuychepxFtj3wWYa/client_golang/prometheus.InstrumentHandlerFuncWithOpts.func1(0x11396e0, 0xc422f7e1a0, 0xc421eb3860)
	/home/kubuxu/AUR/go-ipfs-git/src/src/gx/ipfs/QmR3KwhXCRLTNZB59vELb2HhEWrGy9nuychepxFtj3wWYa/client_golang/prometheus/http.go:287 +0x2ab
net/http.HandlerFunc.ServeHTTP(0xc4220503c0, 0x11396e0, 0xc422f7e1a0, 0xc421eb3860)
	/usr/lib/go/src/net/http/server.go:1726 +0x44
net/http.(*ServeMux).ServeHTTP(0xc421fff800, 0x11396e0, 0xc422f7e1a0, 0xc421eb3860)
	/usr/lib/go/src/net/http/server.go:2022 +0x7f
net/http.serverHandler.ServeHTTP(0xc422007e00, 0x11396e0, 0xc422f7e1a0, 0xc421eb3860)
	/usr/lib/go/src/net/http/server.go:2202 +0x7d
net/http.(*conn).serve(0xc422e34680, 0x113a4e0, 0xc422f31ac0)
	/usr/lib/go/src/net/http/server.go:1579 +0x4b7
created by net/http.(*Server).Serve
	/usr/lib/go/src/net/http/server.go:2293 +0x44d
```

Replicate
```
 echo dcf | ipfs add --raw-leaves               
added zdvgqGt3bciGPAzEbMDDqZKPmJxjCANeRwZUQWHb5hsXgw7Ba zdvgqGt3bciGPAzEbMDDqZKPmJxjCANeRwZUQWHb5hsXgw7Ba
 curl -s localhost:8081/ipfs/zdvgqGt3bciGPAzEbMDDqZKPmJxjCANeRwZUQWHb5hsXgw7Ba

```

### PR DESCRIPTION
**Pull request by Kubuxu**
This doesn't resolve issue yet, I have to refactor a bit so and I want to run CI over this refactor.

Resolves #3399

### CODE REVIEW

**Comment by Kubuxu** [comment]
Ooops, I broke tests.

**Comment by Kubuxu** [comment]
Should be fixed. Let's see CI run.

File: unixfs/io/bufdagreader.go [hunk 1]
```diff
@@ -0,0 +1,41 @@
+package io
+
+import (
+	"bytes"
+	"context"
+	"io"
+)
+
+type bufDagReader struct {
```
Comment (whyrusleeping) [inline unixfs/io/bufdagreader.go]: Why don't we go ahead and replace all of the `NewRNSC` stuff with this now? It seems weird that they both serve roughly the same purpose.

File: unixfs/io/pbdagreader.go [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+		return io.EOF
+	}
+
+	nxt, err := dr.promises[dr.linkPosition].Get(ctx)
+	if err != nil {
+		return err
+	}
+	dr.linkPosition++
+
+	switch nxt := nxt.(type) {
```
Comment (whyrusleeping) [inline unixfs/io/pbdagreader.go]: This should just call `NewDagReader` now, right?

Comment (Kubuxu) [inline unixfs/io/pbdagreader.go]: Not really, as cases for Dir, metadata and symlinks are a bit different.

File: unixfs/io/bufdagreader.go (hunk 1 above)
Comment (Kubuxu) [inline unixfs/io/bufdagreader.go]: Yeah, the interface of DagReader supersedes RSC and bufDagReader is nop-closers on bytes.

File: unixfs/io/pbdagreader.go (hunk 2 above)
Comment (Kubuxu) [inline unixfs/io/pbdagreader.go]: Handled it a bit differently.

Comment (whyrusleeping) [inline unixfs/io/pbdagreader.go]: Are you sure? They look like the same piece of code...

**Pull request merged**